Add license check for FIPS (#181187)
elastic/kibana: docs/user/security/fips-140-2.asciidoc

[role="xpack"]
[[xpack-security-fips-140-2]]

> === FIPS 140-2

Author: Heya @georgewallace ! Would you or someone on you team be able to take a look at this documentation?

Author: Heya @elastic/kibana-docs Would someone mind taking a look?

Reviewer: To make this page appear in the output, you'll need to add `include::fips-140-2.asciidoc[]` at the end of the `docs/user/security/index.asciidoc` file (or in some other navigation file, depending on where it ought to appear in the navigation tree).

Reviewer: I've added the page to the navigation in c37004ac14ef151b52f138a1fe649f93176fdee1.  I think it could honestly go in either of these sections: https://www.elastic.co/guide/en/kibana/current/using-kibana-with-security.html or https://www.elastic.co/guide/en/kibana/current/xpack-security.html.  I've added it to the latter for now but can easily move if the former makes more sense for this content.

experimental[] The Federal Information Processing Standard (FIPS) Publication 140-2, (FIPS PUB 140-2), 
titled "Security Requirements for Cryptographic Modules" is a U.S. government computer security standard
used to approve cryptographic modules.

{kib} offers a FIPS 140-2 compliant mode and as such can run in a Node.js environment configured with a FIPS 
140-2 compliant OpenSSL3 provider.

> To run {kib} in FIPS mode, your cluster must be have a Platinum or Enterprise license.

Reviewer: ```suggestion
To run {kib} in FIPS mode, you must have the appropriate {subscriptions}[subscription].
```

Reviewer: Fixed in c37004ac14ef151b52f138a1fe649f93176fdee1

[IMPORTANT]
============================================================================
The Node bundled with {kib} is not configured for FIPS 140-2. You must configure a FIPS 140-2 compliant OpenSSL3 

> provider. Please consult the Node.js documentation to learn how to configure your environment.

Reviewer: ```suggestion
provider. Consult the Node.js documentation to learn how to configure your environment.
```

Reviewer: Fixed in c37004ac14ef151b52f138a1fe649f93176fdee1

============================================================================

For {kib}, adherence to FIPS 140-2 is ensured by:

* Using FIPS approved / NIST recommended cryptographic algorithms.

* Delegating the implementation of these cryptographic algorithms to a NIST validated cryptographic module
(available via Node.js configured with an OpenSSL3 provider).

* Allowing the configuration of {kib} in a FIPS 140-2 compliant manner, as documented below. 

> ==== Configuring {kib} for FIPS 140-2

Reviewer 2: Do we need to do anything about https://github.com/elastic/kibana/blob/975eeed255d0413f57827d830f991cdf66bdf788/config/node.options#L14-L15

Author: Good question 😅 Im not sure how that conflicts in a FIPS setup. I can do some investigation.

Is there anyway the KB admin could override this currently?

Reviewer 2: It can be removed, Kibana archives will have this at `$KIBANA_HOME/config/node.options`. or for deb/rpm `/etc/kibana/node.options`.

Reviewer 3: I _think_ it's ok to leave. You can use multiple providers alongside the FIPS provider. In #183777 I ran the FIPS agent pipeline appending FIPS options to the existing `node.options` without issue. That said, not sure it's "FIPS compliant" and still worth double checking @kc13greiner.

Author: Ok, we should be able to leave it:

When I configure node with:
```
[provider_sect]
default = default_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
# fips = fips_sect
legacy = legacy_sect
```

It runs with the legacy provider (w/o FIPS)  and allows `md4`/`md5`

<img width="1446" alt="Screenshot 2024-06-11 at 5 25 10 PM" src="https://github.com/elastic/kibana/assets/21210601/b68b2557-a877-4e65-b087-f9aad4c4b95d">

If I include `fips` as a provider, unsupported algorithms are not allowed and it doesn't even mention legacy mode anymore

```
[provider_sect]
default = default_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
fips = fips_sect
legacy = legacy_sect
```
<img width="1806" alt="Screenshot 2024-06-11 at 5 26 56 PM" src="https://github.com/elastic/kibana/assets/21210601/e6333090-4080-4859-9849-85d3c4dd0357">

Apart from setting `xpack.security.experimental.fipsMode.enabled` to `true` in your {kib} config, a number of security related 
settings need to be reviewed and configured in order to run {kib} successfully in a FIPS 140-2 compliant Node.js 
environment.

===== Kibana keystore

FIPS 140-2 (via NIST Special Publication 800-132) dictates that encryption keys should at least have an effective 
strength of 112 bits. As such, the Kibana keystore that stores the application’s secure settings needs to be 
password protected with a password that satisfies this requirement. This means that the password needs to be 14 bytes 
long which is equivalent to a 14 character ASCII encoded password, or a 7 character UTF-8 encoded password.

> For more information on how to set this password, please see the <<change-password,keystore documentation>>.

Reviewer: ```suggestion
For more information on how to set this password, refer to <<change-password,keystore documentation>>.
```

Reviewer: Fixed in c37004ac14ef151b52f138a1fe649f93176fdee1
